File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/fieldstudy/FieldStudyService.kt
package pt.isel.odin.service.fieldstudy

import jakarta.transaction.Transactional
import org.springframework.stereotype.Service
import pt.isel.odin.http.controllers.fieldstudy.models.SaveFieldStudyInputModel
import pt.isel.odin.http.controllers.fieldstudy.models.UpdateFieldStudyInputModel
import pt.isel.odin.model.FieldStudy
import pt.isel.odin.repository.DepartmentRepository
import pt.isel.odin.repository.FieldStudyRepository
import pt.isel.odin.service.fieldstudy.error.DeleteFieldStudyError
import pt.isel.odin.service.fieldstudy.error.GetFieldStudyError
import pt.isel.odin.service.fieldstudy.error.SaveUpdateFieldStudyError
import pt.isel.odin.utils.failure
import pt.isel.odin.utils.success

/**
 * Service for Fields Of Study.
 */
@Service
class FieldStudyService(
    private val fieldStudyRepository: FieldStudyRepository,
    private val departmentRepository: DepartmentRepository
) {

    /**
     * Gets a field of study by its id.
     *
     * @param id the field of study id
     *
     * @return the [GetFieldStudyResult] if found, [GetFieldStudyError.NotFoundFieldStudy] otherwise
     */
    fun getById(id: Long): GetFieldStudyResult =
        fieldStudyRepository.findById(id)
            .map<GetFieldStudyResult> { field -> success(field) }
            .orElse(failure(GetFieldStudyError.NotFoundFieldStudy))

    /**
     * Gets all fields of study.
     *
     * @return the [GetAllFieldsStudyResult] with the list of [FieldStudy]
     */
    fun getAll(): GetAllFieldsStudyResult = success(fieldStudyRepository.findAll())

    /**
     * Saves a field of study.
     *
     * @param saveFieldStudyInputModel the field of study to save
     *
     * @return the [CreationFieldStudyResult] if saved, [SaveUpdateFieldStudyError] otherwise
     */
    @Transactional
    fun save(saveFieldStudyInputModel: SaveFieldStudyInputModel): CreationFieldStudyResult {
        if (fieldStudyRepository.findByName(saveFieldStudyInputModel.name).isPresent) {
            return failure(SaveUpdateFieldStudyError.AlreadyExistsFieldStudy)
        }

        val department = departmentRepository.findById(saveFieldStudyInputModel.department)
        if (department.isEmpty) return failure(SaveUpdateFieldStudyError.NotFoundDepartment)

        if (saveFieldStudyInputModel.name.isBlank()) {
            return failure(SaveUpdateFieldStudyError.IncorrectNameFieldStudy)
        }

        return success(fieldStudyRepository.save(saveFieldStudyInputModel.toFieldStudy(department.get())))
    }

    /**
     * Updates a field of study.
     *
     * @param updateFieldStudyInputModel the field of study to update
     *
     * @return the [CreationFieldStudyResult] if updated, [SaveUpdateFieldStudyError] otherwise
     */
    @Transactional
    fun update(updateFieldStudyInputModel: UpdateFieldStudyInputModel): CreationFieldStudyResult {
        val department = departmentRepository.findById(updateFieldStudyInputModel.department)
        if (department.isEmpty) return failure(SaveUpdateFieldStudyError.NotFoundDepartment)

        if (updateFieldStudyInputModel.name.isBlank()) {
            return failure(SaveUpdateFieldStudyError.IncorrectNameFieldStudy)
        }

        return fieldStudyRepository.findById(updateFieldStudyInputModel.id)
            .map<CreationFieldStudyResult> { field ->
                success(
                    fieldStudyRepository.save(
                        field.copy(
                            name = updateFieldStudyInputModel.name,
                            department = department.get()
                        )
                    )
                )
            }.orElse(failure(SaveUpdateFieldStudyError.NotFoundFieldStudy))
    }

    /**
     * Deletes a field of study by its id.
     *
     * @param id the field of study id
     *
     * @return the [DeleteFieldStudyResult] if deleted, [DeleteFieldStudyError] otherwise
     */
    @Transactional
    fun delete(id: Long): DeleteFieldStudyResult =
        fieldStudyRepository.findById(id)
            .map<DeleteFieldStudyResult> { field ->
                fieldStudyRepository.delete(field)
                success(field)
            }.orElse(failure(DeleteFieldStudyError.NotFoundFieldStudy))
}

File: code/frontend/Tyr/src/services/fieldstudy/FieldStudyService.ts
import {apiRequest} from '../../axios/apiRequest';
import {
    CreationFieldStudyResult,
    DeleteFieldStudyResult,
    GetAllFieldsStudyResult,
    GetFieldStudyResult
} from "./FieldStudyResult";
import {FieldStudyInputModel} from "./models/FieldStudyInputModel";

/**
 * Service to manage field studies
 */
export namespace FieldStudyService {

    const basePath = '/fieldsstudy';

    /**
     * Get field of study by id
     *
     * @param id the id of the field of study
     *
     * @returns the field of study
     */
    export async function getById(id: number): Promise<GetFieldStudyResult> {
        return apiRequest('get', `${basePath}/${id}`);
    }

    /**
     * Get all fields of study
     *
     * @returns all fields of study
     */
    export async function getAll(): Promise<GetAllFieldsStudyResult> {
        return apiRequest('get', basePath);
    }

    /**
     * Save a field of study
     *
     * @param fieldStudyRequest the field of study to save
     *
     * @returns the result of the creation
     */
    export async function save(fieldStudyRequest: FieldStudyInputModel): Promise<CreationFieldStudyResult> {
        return apiRequest('post', `${basePath}/save`, fieldStudyRequest);
    }

    /**
     * Update a field of study
     *
     * @param fieldStudyRequest the field of study to update
     *
     * @returns the result of the update
     */
    export async function update(fieldStudyRequest: FieldStudyInputModel): Promise<CreationFieldStudyResult> {
        return apiRequest('put', `${basePath}/update`, fieldStudyRequest);
    }

    /**
     * Delete a field of study by id
     *
     * @param id the id of the field of study
     *
     * @returns the result of the deletion
     */
    export async function deleteById(id: number): Promise<DeleteFieldStudyResult> {
        return apiRequest('delete', `${basePath}/${id}`);
    }
}
